design/bridge_design.md

```diff
@@ -0,0 +1,364 @@
+) public returns (bool)
+```
+
+Internally, the function will hash the message and compare it with the result of the key in `receivedMessages` map. If the map contains a `true` under this key, then the message has been received by this contract and `verifyMessageReceived` will return true. This is useful as a cheap-ish way to verify something has happened by utilizing the end user's client (or browser) as the transfer mechanism for the full message.  
+
+
+
+### Submit Out Of Network Messages
+
+This is the smart contract function which is used to store messages sent from the other linked layer. The function will be called by the `ManagementContract` on L1 and the `enclave` on L2. It should be access controlled and called according to the `consistencyLevel` and Obscuro platform rules.
+
+```solidity 
+function submitOutOfNetworkMessage(
+        address      sender,
+        uint??       sequence,
+        uint32       nonce,
+        bytes memory topic,
+        bytes memory payload
+) public onlyRole(ADMINISTRATOR|OWNER)
+```
+
+This function should not be callable by any users or contracts unless they have the `ADMINISTRATOR` role or are the `Owner` of the `MessageBus` contract. Along with those requirements, on the Obscuro layer 2, there is an additional security measure - Any incoming transaction to this function is blocked, even if it is coming from an address that has the correct role. The protocol will block transactions that do not originate from inside of the enclave due to consuming L1 blocks. This means that even if the enclave gets hacked and the keys leak the verifiers will block withdraws.
+
+When called, the function should store the message indexed in storage - map of `senders` that contains a map of `topics` which points to array of `messages` should be sufficient. 
+Along with it, the hash of the whole message should be used as a key to store a message received flag inside the map `receivedMessages`. This map enables quick verification that a message is valid and received.
+
+### Query Messages
+
+This function allows querying messages on a topic, by a certain user. 
+
+```solidity 
+function queryMessages(
+        address      sender,
+        bytes memory topic,
+        uint256      fromIndex,
+        uint256      toIndex,
+) public returns (bytes [] memory)
+```
+
+This function should return all the messages between the requested `fromIndex` and `toIndex` of the array storing the submitted messages. The array can be found when going through the map of `senders` under the concrete `topic`. It is a more expensive way to get messages on-chain without relying on users to ferry them around. 
+
+### MessageBus Internal workings - L1 to L2
+
+When a block from `L1` is processed by the `enclave` and transactions inside of it result in `MessagePublished` events, the `enclave` will submit them as messages to the inbox of the `L2` contract. This enables the `L2 smart contracts` to use/consume messages from `L1`. 
```

Comment:
`transactions inside of it result in MessagePublished events` are these actual solidity events or are they "events" in the sense of something that happened ?

Can this be made clearer ?